### PR TITLE
docs(appium): add plugin development guide

### PR DESCRIPTION
**Pull request by jlipps**

### CODE REVIEW

**Comment by jlipps** [comment]
cc @appium/documentors

File: packages/appium/docs/en/ecosystem/build-plugins.md [hunk 1]
```diff
@@ -2,4 +2,436 @@
...
+### Extend Appium's `BasePlugin` class
+
+Ultimately, your plugin is much easier to write because most of the hard work of defining patterns
+for overriding commands is done for you. This is
+all encoded up as a class which Appium exports for you to use, called `BasePlugin`. It is exported
+from `appium/plugin`, so you can use one of these styles to import it and create your *own* class
+that extends it:
+
+```js
+import {BasePlugin} from 'appium/plugin';
```
Comment (mykola-mokhnach) [inline packages/appium/docs/en/ecosystem/build-plugins.md:76]: I assume we should also mention the base plugin must be imported as a peer dependency for this to work?

File: packages/appium/docs/en/ecosystem/build-plugins.md [hunk 2]
```diff
@@ -2,4 +2,436 @@
...
+Now, you can run Appium locally (`npm exec appium` or `npx appium`), and because your plugin is
+listed as a dependency alongside it, it will be automatically "installed" and available. You can
+design your e2e tests this way, or if you're writing them in Node.js, you can simply import
+Appium's start server methods to handle starting and stopping the Appium server in Node.
+
+Another way to do local development with an existing Appium server install is to simply install
+your plugin locally:
+
+```
+appium plugin install --source=local /path/to/your/plugin/dev/dir
```
Comment (mykola-mokhnach) [inline packages/appium/docs/en/ecosystem/build-plugins.md:138]: isn't this repeating the previous paragraph?

File: packages/appium/docs/en/ecosystem/build-plugins.md [hunk 3]
```diff
@@ -2,4 +2,436 @@
...
+examples can be found at the relevant client docs).
+
+An alternative to this way of doing things is to overload a command which all WebDriver clients
+have access to already: Execute Script. Make sure to read the section on [adding new
+commands](./build-drivers.md#extend-the-existing-protocol-with-new-commands) in
+the Building Drivers guide to understand the way this works in general. The way it works with
+plugins is only slightly different. Let's look at an example taken from Appium's `fake-plugin`:
+
+```js
+static executeMethodMap = {
```
Comment (mykola-mokhnach) [inline packages/appium/docs/en/ecosystem/build-plugins.md:372]: from this example it is unclear the definition should be located inside of the plugin class

File: packages/appium/docs/en/ecosystem/build-plugins.md (hunk 1 above)
Comment (jlipps) [reply packages/appium/docs/en/ecosystem/build-plugins.md:76]: it's appium that must be imported as a peer dependency, which is already mentioned above

File: packages/appium/docs/en/ecosystem/build-plugins.md (hunk 2 above)
Comment (jlipps) [reply packages/appium/docs/en/ecosystem/build-plugins.md:138]: yeah, will remove.

File: packages/appium/docs/en/ecosystem/build-plugins.md (hunk 3 above)
Comment (jlipps) [reply packages/appium/docs/en/ecosystem/build-plugins.md:372]: ok made that more clear.

File: packages/appium/docs/en/ecosystem/build-plugins.md [hunk 4]
```diff
@@ -2,4 +2,435 @@
...
+### Update the Appium server object
+
+You probably don't normally need to update the Appium server object (which is an Express server
+having already been configured in a variety of ways). But, for example, you could add new Express
+middleware to the server to support your plugin's requirements. To update the server you must
+implement the `static async updateServer` method in your class. This method takes two parameters:
+
+* `expressApp`: the Express app object
+* `httpServer`: the Node HTTP server object
+
```
Comment (saikrishna321) [inline packages/appium/docs/en/ecosystem/build-plugins.md:409]: @jlipps update server also takes third argument which is cliArgs.

Comment (jlipps) [reply packages/appium/docs/en/ecosystem/build-plugins.md:409]: good catch, thanks

File: packages/appium/docs/en/ecosystem/build-plugins.md [hunk 5]
```diff
@@ -2,4 +2,436 @@
...
+first two arguments are `next` and `driver`, followed by the script name and args. `BasePlugin`
+implements a helper method which we can simply call with all of these arguments.
+
+Overriding Execute Methods from drivers works as you'd expect: if your plugin defines an Execute
+Method with the same name as that of a driver, your command (in this case `plugMeIn`) will be
+called first. You can choose to run the driver's original behaviour via `next` if you want.
+
+### Update the Appium server object
+
+You probably don't normally need to update the Appium server object (which is an Express server
```
Comment (mykola-mokhnach) [inline packages/appium/docs/en/ecosystem/build-plugins.md:402]: we probably need to mention where the Express server class is located in the base driver

File: packages/appium/docs/en/ecosystem/build-plugins.md [hunk 6]
```diff
@@ -2,4 +2,436 @@
...
+naturally do this by adding a handler for `deleteSession`. This works in most cases, except when
+the session does not finish cleanly. Appium sometimes determines that a session has finished
+unexpectedly, and in these situations, Appium will look for a method called `onUnexpectedShutdown`
+in your plugin class, which will be called (passing the current session driver as the first
+parameter, and the error object representing the cause of the shutdown as the second), giving you
+an opportunity to take any steps that might be necessary to clean up from the session. For example,
+keeping in mind that the function is not `await`ed you could implement something like this:
+
+```js
+async onUnexpectedShutdown(driver, cause) {
```
Comment (mykola-mokhnach) [inline packages/appium/docs/en/ecosystem/build-plugins.md:430]: I assume other examples should also be more explicit about these methods should not be global but rather plugin methods

**Review by mykola-mokhnach — APPROVED** [review]

**Pull request merged**